### PR TITLE
add model archiving

### PR DESCRIPTION
**Pull request by joelgrus**
at a high level, I added a `model.archive()` method that saves vocabulary, best weights, and the training config to a .tar.gz file. and then I added a corresponding `Model.from_archive()` to re-instantiate the model.

I also modified the `train` command to automatically archive the results. As it stands right now, this will crash if the config doesn't specify a validation dataset, since in that case `best.th` never gets written out. I'm not sure what the right behavior is there? Should training fail immediately ("validation dataset is required")? Should the archive failover to the weights from the last epoch? Something else? Ideas welcome.

I tried really hard to instantiate the model directly from the tar file, but I couldn't get `torch.load()` to work with `tarfile.extractfile()`, so I ended up having to untar the archive to a temp directory, load the model, and then wipe out the temp directory. Not elegant, but it actually made the code simpler because I could just use the existing `from_files()` functionality.

### CODE REVIEW

File: allennlp/models/model.py [hunk 1]
```diff
@@ -87,8 +101,70 @@ def get_metrics(self, reset: bool = False) -> Dict[str, float]:
...
+        """
+        Instantiates a model from an archived `tar.gz` file.
+
+        Parameters
+        ----------
+        archive_file: ``str``
+            The archive file to load the model from.
+        cuda_device: ``int``, optional (default = -1)
+            If `cuda_device` is >= 0, the model will be loaded onto the
+            corresponding GPU. Otherwise it will be loaded to CPU.
```
Comment (DeNeutoy) [inline allennlp/models/model.py]: "loaded onto the cpu"

File: allennlp/models/model.py [hunk 2]
```diff
@@ -87,8 +101,70 @@ def get_metrics(self, reset: bool = False) -> Dict[str, float]:
         # pylint: disable=unused-argument,no-self-use
         return {}
 
+    def archive(self,                                       # pylint: disable=no-self-use
```
Comment (DeNeutoy) [inline allennlp/models/model.py]: What do you think about these be separate functions in a `model_utils.py` or something? A research user doesn't need to know about this and it unnecessarily clutters the `Model` API.

File: tests/models/model_test.py [hunk 3]
```diff
@@ -0,0 +1,73 @@
+# pylint: disable=invalid-name
+import os
+import json
+
+from allennlp.common.testing import AllenNlpTestCase
+from allennlp.commands.train import train_model, _CONFIG_FILE_KEY
+from allennlp.models.model import Model
+
+import torch
```
Comment (DeNeutoy) [inline tests/models/model_test.py]: nit: can you move this import above the allennlp ones?

File: tests/models/model_test.py [hunk 4]
```diff
@@ -0,0 +1,73 @@
...
+
+class ModelTest(AllenNlpTestCase):
+    def test_archiving(self):
+        super(ModelTest, self).setUp()
+
+        print(self.TEST_DIR)
```
Comment (DeNeutoy) [inline tests/models/model_test.py]: Remove this.

File: allennlp/models/model.py (hunk 2 above)
Comment (joelgrus) [inline allennlp/models/model.py]: yeah, I considered putting them in `models/archive.py`, my only rationale against it is that archiving functions might need access to internal model state, but for now (when we're just archiving based on what's on disk) I guess we don't. I'll move them out

**Comment by DeNeutoy** [comment]
Also on the saving the best model front, could we just do nothing if `best.th` is not there? I feel like that's the most straightforward approach until we need something better.

**Pull request merged**